// @codingame/monaco-vscode-b1e8558d-1726-5299-bc75-e43ee6d1a124-common/vscode/src/vs/workbench/browser/parts/editor/binaryEditor.js

import { __decorate, __param } from '@codingame/monaco-vscode-api/external/tslib/tslib.es6';
import { localize } from '@codingame/monaco-vscode-api/vscode/vs/nls';
import { Emitter } from '@codingame/monaco-vscode-api/vscode/vs/base/common/event';
import { BinaryEditorModel } from '@codingame/monaco-vscode-c4e2825a-b5b1-5f0e-b547-068c32e06d50-common/vscode/vs/workbench/common/editor/binaryEditorModel';
import { IStorageService } from '@codingame/monaco-vscode-api/vscode/vs/platform/storage/common/storage.service';
import { ByteSize } from '@codingame/monaco-vscode-api/vscode/vs/platform/files/common/files';
import { EditorPlaceholder } from '@codingame/monaco-vscode-ad89fae6-94f0-5ac2-a185-22dea4b68ee0-common/vscode/vs/workbench/browser/parts/editor/editorPlaceholder';

let BaseBinaryResourceEditor = class BaseBinaryResourceEditor extends EditorPlaceholder {
    constructor(id, group, callbacks, telemetryService, themeService, storageService) {
        super(id, group, telemetryService, themeService, storageService);
        this.callbacks = callbacks;
        this._onDidChangeMetadata = this._register(( new Emitter()));
        this.onDidChangeMetadata = this._onDidChangeMetadata.event;
        this._onDidOpenInPlace = this._register(( new Emitter()));
        this.onDidOpenInPlace = this._onDidOpenInPlace.event;
    }
    getTitle() {
        return this.input ? this.input.getName() : ( localize(2821, "Binary Viewer"));
    }
    async getContents(input, options) {
        const model = await input.resolve();
        if (!(model instanceof BinaryEditorModel)) {
            throw ( new Error('Unable to open file as binary'));
        }
        const size = model.getSize();
        this.handleMetadataChanged(typeof size === 'number' ? ByteSize.formatSize(size) : '');
        return {
            icon: '$(warning)',
            label: ( localize(
                2822,
                "The file is not displayed in the text editor because it is either binary or uses an unsupported text encoding."
            )),
            actions: [
                {
                    label: ( localize(2823, "Open Anyway")),
                    run: async () => {
                        await this.callbacks.openInternal(input, options);
                        this._onDidOpenInPlace.fire();
                    }
                }
            ]
        };
    }
    handleMetadataChanged(meta) {
        this.metadata = meta;
        this._onDidChangeMetadata.fire();
    }
    getMetadata() {
        return this.metadata;
    }
};
BaseBinaryResourceEditor = ( __decorate([
    ( __param(5, IStorageService))
], BaseBinaryResourceEditor));

export { BaseBinaryResourceEditor };
