// @codingame/monaco-vscode-9a6d8b6c-ad4c-5ea3-9558-f43d6dc4c0ad-common/vscode/src/vs/editor/common/codecs/linesCodec/tokens/carriageReturn.js

import { BaseToken } from '../../baseToken.js';
import { Range } from '@codingame/monaco-vscode-api/vscode/vs/editor/common/core/range';
import { Position } from '@codingame/monaco-vscode-api/vscode/vs/editor/common/core/position';
import { VSBuffer } from '@codingame/monaco-vscode-api/vscode/vs/base/common/buffer';

class CarriageReturn extends BaseToken {
    static { this.symbol = '\r'; }
    static { this.byte = VSBuffer.fromString(CarriageReturn.symbol); }
    get byte() {
        return CarriageReturn.byte;
    }
    get text() {
        return CarriageReturn.symbol;
    }
    static newOnLine(line, atColumnNumber) {
        const { range } = line;
        const startPosition = ( new Position(range.startLineNumber, atColumnNumber));
        const endPosition = ( new Position(range.startLineNumber, atColumnNumber + this.symbol.length));
        return ( new CarriageReturn(Range.fromPositions(startPosition, endPosition)));
    }
    toString() {
        return `carriage-return${this.range}`;
    }
}

export { CarriageReturn };
