// @codingame/monaco-vscode-7bbc9e7d-eeae-55fc-8bf9-dc2f66e0dc73-common/vscode/src/vs/workbench/contrib/terminalContrib/zoom/common/terminal.zoom.js

import { isMacintosh } from '@codingame/monaco-vscode-api/vscode/vs/base/common/platform';
import { localize } from '@codingame/monaco-vscode-api/vscode/vs/nls';

var TerminalZoomCommandId;
(function (TerminalZoomCommandId) {
    TerminalZoomCommandId["FontZoomIn"] = "workbench.action.terminal.fontZoomIn";
    TerminalZoomCommandId["FontZoomOut"] = "workbench.action.terminal.fontZoomOut";
    TerminalZoomCommandId["FontZoomReset"] = "workbench.action.terminal.fontZoomReset";
})(TerminalZoomCommandId || (TerminalZoomCommandId = {}));
var TerminalZoomSettingId;
(function (TerminalZoomSettingId) {
    TerminalZoomSettingId["MouseWheelZoom"] = "terminal.integrated.mouseWheelZoom";
})(TerminalZoomSettingId || (TerminalZoomSettingId = {}));
const terminalZoomConfiguration = {
    [TerminalZoomSettingId.MouseWheelZoom]: {
        markdownDescription: isMacintosh
            ? ( localize(
            10319,
            "Zoom the font of the terminal when using mouse wheel and holding `Cmd`."
        ))
            : ( localize(
            10320,
            "Zoom the font of the terminal when using mouse wheel and holding `Ctrl`."
        )),
        type: 'boolean',
        default: false
    },
};

export { TerminalZoomCommandId, TerminalZoomSettingId, terminalZoomConfiguration };
