// @codingame/monaco-vscode-9a6d8b6c-ad4c-5ea3-9558-f43d6dc4c0ad-common/vscode/src/vs/editor/common/codecs/simpleCodec/tokens/brackets.js

import { BaseToken } from '../../baseToken.js';
import { Range } from '@codingame/monaco-vscode-api/vscode/vs/editor/common/core/range';
import { Position } from '@codingame/monaco-vscode-api/vscode/vs/editor/common/core/position';

class LeftBracket extends BaseToken {
    static { this.symbol = '['; }
    get text() {
        return LeftBracket.symbol;
    }
    static newOnLine(line, atColumnNumber) {
        const { range } = line;
        const startPosition = ( new Position(range.startLineNumber, atColumnNumber));
        const endPosition = ( new Position(range.startLineNumber, atColumnNumber + this.symbol.length));
        return ( new LeftBracket(Range.fromPositions(startPosition, endPosition)));
    }
    toString() {
        return `left-bracket${this.range}`;
    }
}
class RightBracket extends BaseToken {
    static { this.symbol = ']'; }
    get text() {
        return RightBracket.symbol;
    }
    static newOnLine(line, atColumnNumber) {
        const { range } = line;
        const startPosition = ( new Position(range.startLineNumber, atColumnNumber));
        const endPosition = ( new Position(range.startLineNumber, atColumnNumber + this.symbol.length));
        return ( new RightBracket(Range.fromPositions(startPosition, endPosition)));
    }
    toString() {
        return `right-bracket${this.range}`;
    }
}

export { LeftBracket, RightBracket };
