// @codingame/monaco-vscode-9a6d8b6c-ad4c-5ea3-9558-f43d6dc4c0ad-common/vscode/src/vs/editor/common/codecs/simpleCodec/tokens/parentheses.js

import { BaseToken } from '../../baseToken.js';
import { Range } from '@codingame/monaco-vscode-api/vscode/vs/editor/common/core/range';
import { Position } from '@codingame/monaco-vscode-api/vscode/vs/editor/common/core/position';

class LeftParenthesis extends BaseToken {
    static { this.symbol = '('; }
    get text() {
        return LeftParenthesis.symbol;
    }
    static newOnLine(line, atColumnNumber) {
        const { range } = line;
        const startPosition = ( new Position(range.startLineNumber, atColumnNumber));
        const endPosition = ( new Position(range.startLineNumber, atColumnNumber + this.symbol.length));
        return ( new LeftParenthesis(Range.fromPositions(startPosition, endPosition)));
    }
    toString() {
        return `left-parenthesis${this.range}`;
    }
}
class RightParenthesis extends BaseToken {
    static { this.symbol = ')'; }
    get text() {
        return RightParenthesis.symbol;
    }
    static newOnLine(line, atColumnNumber) {
        const { range } = line;
        const startPosition = ( new Position(range.startLineNumber, atColumnNumber));
        const endPosition = ( new Position(range.startLineNumber, atColumnNumber + this.symbol.length));
        return ( new RightParenthesis(Range.fromPositions(startPosition, endPosition)));
    }
    toString() {
        return `right-parenthesis${this.range}`;
    }
}

export { LeftParenthesis, RightParenthesis };
