// @codingame/monaco-vscode-7bbc9e7d-eeae-55fc-8bf9-dc2f66e0dc73-common/vscode/src/vs/workbench/contrib/terminalContrib/suggest/common/terminal.suggest.js


var TerminalSuggestCommandId;
(function (TerminalSuggestCommandId) {
    TerminalSuggestCommandId["SelectPrevSuggestion"] = "workbench.action.terminal.selectPrevSuggestion";
    TerminalSuggestCommandId["SelectPrevPageSuggestion"] = "workbench.action.terminal.selectPrevPageSuggestion";
    TerminalSuggestCommandId["SelectNextSuggestion"] = "workbench.action.terminal.selectNextSuggestion";
    TerminalSuggestCommandId["SelectNextPageSuggestion"] = "workbench.action.terminal.selectNextPageSuggestion";
    TerminalSuggestCommandId["AcceptSelectedSuggestion"] = "workbench.action.terminal.acceptSelectedSuggestion";
    TerminalSuggestCommandId["AcceptSelectedSuggestionEnter"] = "workbench.action.terminal.acceptSelectedSuggestionEnter";
    TerminalSuggestCommandId["HideSuggestWidget"] = "workbench.action.terminal.hideSuggestWidget";
    TerminalSuggestCommandId["ClearSuggestCache"] = "workbench.action.terminal.clearSuggestCache";
    TerminalSuggestCommandId["RequestCompletions"] = "workbench.action.terminal.requestCompletions";
    TerminalSuggestCommandId["ResetWidgetSize"] = "workbench.action.terminal.resetSuggestWidgetSize";
    TerminalSuggestCommandId["ToggleDetails"] = "workbench.action.terminal.suggestToggleDetails";
    TerminalSuggestCommandId["ToggleDetailsFocus"] = "workbench.action.terminal.suggestToggleDetailsFocus";
    TerminalSuggestCommandId["ConfigureSettings"] = "workbench.action.terminal.configureSuggestSettings";
})(TerminalSuggestCommandId || (TerminalSuggestCommandId = {}));
const defaultTerminalSuggestCommandsToSkipShell = [
    TerminalSuggestCommandId.SelectPrevSuggestion,
    TerminalSuggestCommandId.SelectPrevPageSuggestion,
    TerminalSuggestCommandId.SelectNextSuggestion,
    TerminalSuggestCommandId.SelectNextPageSuggestion,
    TerminalSuggestCommandId.AcceptSelectedSuggestion,
    TerminalSuggestCommandId.AcceptSelectedSuggestionEnter,
    TerminalSuggestCommandId.HideSuggestWidget,
    TerminalSuggestCommandId.ClearSuggestCache,
    TerminalSuggestCommandId.RequestCompletions,
    TerminalSuggestCommandId.ToggleDetails,
    TerminalSuggestCommandId.ToggleDetailsFocus,
];

export { TerminalSuggestCommandId, defaultTerminalSuggestCommandsToSkipShell };
