// @codingame/monaco-vscode-9a6d8b6c-ad4c-5ea3-9558-f43d6dc4c0ad-common/vscode/src/vs/workbench/contrib/notebook/browser/diff/notebookDiffEditorBrowser.js

import { RawContextKey } from '@codingame/monaco-vscode-api/vscode/vs/platform/contextkey/common/contextkey';
import { localize } from '@codingame/monaco-vscode-api/vscode/vs/nls';

var DiffSide;
(function (DiffSide) {
    DiffSide[DiffSide["Original"] = 0] = "Original";
    DiffSide[DiffSide["Modified"] = 1] = "Modified";
})(DiffSide || (DiffSide = {}));
const DIFF_CELL_MARGIN = 16;
const NOTEBOOK_DIFF_CELL_INPUT = ( new RawContextKey('notebook.diffEditor.cell.inputChanged', false));
const NOTEBOOK_DIFF_METADATA = ( new RawContextKey('notebook.diffEditor.metadataChanged', false));
const NOTEBOOK_DIFF_CELL_IGNORE_WHITESPACE_KEY = 'notebook.diffEditor.cell.ignoreWhitespace';
const NOTEBOOK_DIFF_CELL_IGNORE_WHITESPACE = ( new RawContextKey(NOTEBOOK_DIFF_CELL_IGNORE_WHITESPACE_KEY, false));
const NOTEBOOK_DIFF_CELL_PROPERTY = ( new RawContextKey('notebook.diffEditor.cell.property.changed', false));
const NOTEBOOK_DIFF_CELL_PROPERTY_EXPANDED = ( new RawContextKey('notebook.diffEditor.cell.property.expanded', false));
const NOTEBOOK_DIFF_CELLS_COLLAPSED = ( new RawContextKey('notebook.diffEditor.allCollapsed', undefined, ( localize(7831, "Whether all cells in notebook diff editor are collapsed"))));
const NOTEBOOK_DIFF_HAS_UNCHANGED_CELLS = ( new RawContextKey('notebook.diffEditor.hasUnchangedCells', undefined, ( localize(7832, "Whether there are unchanged cells in the notebook diff editor"))));
const NOTEBOOK_DIFF_UNCHANGED_CELLS_HIDDEN = ( new RawContextKey('notebook.diffEditor.unchangedCellsAreHidden', undefined, ( localize(7833, "Whether the unchanged cells in the notebook diff editor are hidden"))));
const NOTEBOOK_DIFF_ITEM_KIND = ( new RawContextKey('notebook.diffEditor.item.kind', undefined, ( localize(
 7834,
 "The kind of item in the notebook diff editor, Cell, Metadata or Output"
))));
const NOTEBOOK_DIFF_ITEM_DIFF_STATE = ( new RawContextKey('notebook.diffEditor.item.state', undefined, ( localize(
 7835,
 "The diff state of item in the notebook diff editor, delete, insert, modified or unchanged"
))));

export { DIFF_CELL_MARGIN, DiffSide, NOTEBOOK_DIFF_CELLS_COLLAPSED, NOTEBOOK_DIFF_CELL_IGNORE_WHITESPACE, NOTEBOOK_DIFF_CELL_IGNORE_WHITESPACE_KEY, NOTEBOOK_DIFF_CELL_INPUT, NOTEBOOK_DIFF_CELL_PROPERTY, NOTEBOOK_DIFF_CELL_PROPERTY_EXPANDED, NOTEBOOK_DIFF_HAS_UNCHANGED_CELLS, NOTEBOOK_DIFF_ITEM_DIFF_STATE, NOTEBOOK_DIFF_ITEM_KIND, NOTEBOOK_DIFF_METADATA, NOTEBOOK_DIFF_UNCHANGED_CELLS_HIDDEN };
