// @codingame/monaco-vscode-7bbc9e7d-eeae-55fc-8bf9-dc2f66e0dc73-common/vscode/src/vs/workbench/contrib/terminalContrib/stickyScroll/common/terminalStickyScrollConfiguration.js

import { localize } from '@codingame/monaco-vscode-api/vscode/vs/nls';
import product from '@codingame/monaco-vscode-api/vscode/vs/platform/product/common/product';
import { TerminalSettingId } from '@codingame/monaco-vscode-api/vscode/vs/platform/terminal/common/terminal';

var TerminalStickyScrollSettingId;
(function (TerminalStickyScrollSettingId) {
    TerminalStickyScrollSettingId["Enabled"] = "terminal.integrated.stickyScroll.enabled";
    TerminalStickyScrollSettingId["MaxLineCount"] = "terminal.integrated.stickyScroll.maxLineCount";
})(TerminalStickyScrollSettingId || (TerminalStickyScrollSettingId = {}));
const terminalStickyScrollConfiguration = {
    [TerminalStickyScrollSettingId.Enabled]: {
        markdownDescription: ( localize(
            10266,
            "Shows the current command at the top of the terminal. This feature requires [shell integration]({0}) to be activated. See {1}.",
            'https://code.visualstudio.com/docs/terminal/shell-integration',
            `\`#${TerminalSettingId.ShellIntegrationEnabled}#\``
        )),
        type: 'boolean',
        default: product.quality !== 'stable'
    },
    [TerminalStickyScrollSettingId.MaxLineCount]: {
        markdownDescription: ( localize(
            10267,
            "Defines the maximum number of sticky lines to show. Sticky scroll lines will never exceed 40% of the viewport regardless of this setting."
        )),
        type: 'number',
        default: 5,
        minimum: 1,
        maximum: 10
    },
};

export { TerminalStickyScrollSettingId, terminalStickyScrollConfiguration };
