// @codingame/monaco-vscode-api/vscode/src/vs/workbench/common/theme.js

import { localize } from '../../nls.js';
import { registerColor, transparent, darken, lighten } from '../../platform/theme/common/colorUtils.js';
import { contrastBorder, activeContrastBorder, focusBorder, errorForeground, textLinkForeground } from '../../platform/theme/common/colors/baseColors.js';
import '../../platform/theme/common/colors/chartsColors.js';
import { editorBackground, editorWidgetForeground, editorWidgetBackground, editorInfoForeground, editorForeground, editorWarningForeground, toolbarHoverBackground, widgetBorder, editorErrorForeground } from '../../platform/theme/common/colors/editorColors.js';
import { inputBorder } from '../../platform/theme/common/colors/inputColors.js';
import { treeIndentGuidesStroke, listActiveSelectionBackground, listActiveSelectionForeground } from '../../platform/theme/common/colors/listColors.js';
import '../../platform/theme/common/colors/menuColors.js';
import '../../platform/theme/common/colors/minimapColors.js';
import { scrollbarShadow } from '../../platform/theme/common/colors/miscColors.js';
import '../../platform/theme/common/colors/quickpickColors.js';
import '../../platform/theme/common/colors/searchColors.js';
import { Color } from '../../base/common/color.js';
import { ColorScheme } from '../../platform/theme/common/theme.js';

function WORKBENCH_BACKGROUND(theme) {
    switch (theme.type) {
        case ColorScheme.LIGHT:
            return ( Color.fromHex('#F3F3F3'));
        case ColorScheme.HIGH_CONTRAST_LIGHT:
            return ( Color.fromHex('#FFFFFF'));
        case ColorScheme.HIGH_CONTRAST_DARK:
            return ( Color.fromHex('#000000'));
        default:
            return ( Color.fromHex('#252526'));
    }
}
const TAB_ACTIVE_BACKGROUND = registerColor('tab.activeBackground', editorBackground, ( localize(
    3793,
    "Active tab background color in an active group. Tabs are the containers for editors in the editor area. Multiple tabs can be opened in one editor group. There can be multiple editor groups."
)));
const TAB_UNFOCUSED_ACTIVE_BACKGROUND = registerColor('tab.unfocusedActiveBackground', TAB_ACTIVE_BACKGROUND, ( localize(
    3794,
    "Active tab background color in an unfocused group. Tabs are the containers for editors in the editor area. Multiple tabs can be opened in one editor group. There can be multiple editor groups."
)));
const TAB_INACTIVE_BACKGROUND = registerColor('tab.inactiveBackground', {
    dark: '#2D2D2D',
    light: '#ECECEC',
    hcDark: null,
    hcLight: null,
}, ( localize(
    3795,
    "Inactive tab background color in an active group. Tabs are the containers for editors in the editor area. Multiple tabs can be opened in one editor group. There can be multiple editor groups."
)));
const TAB_UNFOCUSED_INACTIVE_BACKGROUND = registerColor('tab.unfocusedInactiveBackground', TAB_INACTIVE_BACKGROUND, ( localize(
    3796,
    "Inactive tab background color in an unfocused group. Tabs are the containers for editors in the editor area. Multiple tabs can be opened in one editor group. There can be multiple editor groups."
)));
const TAB_ACTIVE_FOREGROUND = registerColor('tab.activeForeground', {
    dark: Color.white,
    light: '#333333',
    hcDark: Color.white,
    hcLight: '#292929'
}, ( localize(
    3797,
    "Active tab foreground color in an active group. Tabs are the containers for editors in the editor area. Multiple tabs can be opened in one editor group. There can be multiple editor groups."
)));
const TAB_INACTIVE_FOREGROUND = registerColor('tab.inactiveForeground', {
    dark: ( transparent(TAB_ACTIVE_FOREGROUND, 0.5)),
    light: ( transparent(TAB_ACTIVE_FOREGROUND, 0.7)),
    hcDark: Color.white,
    hcLight: '#292929'
}, ( localize(
    3798,
    "Inactive tab foreground color in an active group. Tabs are the containers for editors in the editor area. Multiple tabs can be opened in one editor group. There can be multiple editor groups."
)));
const TAB_UNFOCUSED_ACTIVE_FOREGROUND = registerColor('tab.unfocusedActiveForeground', {
    dark: ( transparent(TAB_ACTIVE_FOREGROUND, 0.5)),
    light: ( transparent(TAB_ACTIVE_FOREGROUND, 0.7)),
    hcDark: Color.white,
    hcLight: '#292929'
}, ( localize(
    3799,
    "Active tab foreground color in an unfocused group. Tabs are the containers for editors in the editor area. Multiple tabs can be opened in one editor group. There can be multiple editor groups."
)));
registerColor('tab.unfocusedInactiveForeground', {
    dark: ( transparent(TAB_INACTIVE_FOREGROUND, 0.5)),
    light: ( transparent(TAB_INACTIVE_FOREGROUND, 0.5)),
    hcDark: Color.white,
    hcLight: '#292929'
}, ( localize(
    3800,
    "Inactive tab foreground color in an unfocused group. Tabs are the containers for editors in the editor area. Multiple tabs can be opened in one editor group. There can be multiple editor groups."
)));
const TAB_HOVER_BACKGROUND = registerColor('tab.hoverBackground', null, ( localize(
    3801,
    "Tab background color when hovering. Tabs are the containers for editors in the editor area. Multiple tabs can be opened in one editor group. There can be multiple editor groups."
)));
const TAB_UNFOCUSED_HOVER_BACKGROUND = registerColor('tab.unfocusedHoverBackground', {
    dark: ( transparent(TAB_HOVER_BACKGROUND, 0.5)),
    light: ( transparent(TAB_HOVER_BACKGROUND, 0.7)),
    hcDark: null,
    hcLight: null
}, ( localize(
    3802,
    "Tab background color in an unfocused group when hovering. Tabs are the containers for editors in the editor area. Multiple tabs can be opened in one editor group. There can be multiple editor groups."
)));
const TAB_HOVER_FOREGROUND = registerColor('tab.hoverForeground', null, ( localize(
    3803,
    "Tab foreground color when hovering. Tabs are the containers for editors in the editor area. Multiple tabs can be opened in one editor group. There can be multiple editor groups."
)));
const TAB_UNFOCUSED_HOVER_FOREGROUND = registerColor('tab.unfocusedHoverForeground', {
    dark: ( transparent(TAB_HOVER_FOREGROUND, 0.5)),
    light: ( transparent(TAB_HOVER_FOREGROUND, 0.5)),
    hcDark: null,
    hcLight: null
}, ( localize(
    3804,
    "Tab foreground color in an unfocused group when hovering. Tabs are the containers for editors in the editor area. Multiple tabs can be opened in one editor group. There can be multiple editor groups."
)));
const TAB_BORDER = registerColor('tab.border', {
    dark: '#252526',
    light: '#F3F3F3',
    hcDark: contrastBorder,
    hcLight: contrastBorder,
}, ( localize(
    3805,
    "Border to separate tabs from each other. Tabs are the containers for editors in the editor area. Multiple tabs can be opened in one editor group. There can be multiple editor groups."
)));
const TAB_LAST_PINNED_BORDER = registerColor('tab.lastPinnedBorder', {
    dark: treeIndentGuidesStroke,
    light: treeIndentGuidesStroke,
    hcDark: contrastBorder,
    hcLight: contrastBorder
}, ( localize(
    3806,
    "Border to separate pinned tabs from other tabs. Tabs are the containers for editors in the editor area. Multiple tabs can be opened in one editor group. There can be multiple editor groups."
)));
const TAB_ACTIVE_BORDER = registerColor('tab.activeBorder', null, ( localize(
    3807,
    "Border on the bottom of an active tab. Tabs are the containers for editors in the editor area. Multiple tabs can be opened in one editor group. There can be multiple editor groups."
)));
const TAB_UNFOCUSED_ACTIVE_BORDER = registerColor('tab.unfocusedActiveBorder', {
    dark: ( transparent(TAB_ACTIVE_BORDER, 0.5)),
    light: ( transparent(TAB_ACTIVE_BORDER, 0.7)),
    hcDark: null,
    hcLight: null
}, ( localize(
    3808,
    "Border on the bottom of an active tab in an unfocused group. Tabs are the containers for editors in the editor area. Multiple tabs can be opened in one editor group. There can be multiple editor groups."
)));
const TAB_ACTIVE_BORDER_TOP = registerColor('tab.activeBorderTop', {
    dark: null,
    light: null,
    hcDark: null,
    hcLight: '#B5200D'
}, ( localize(
    3809,
    "Border to the top of an active tab. Tabs are the containers for editors in the editor area. Multiple tabs can be opened in one editor group. There can be multiple editor groups."
)));
const TAB_UNFOCUSED_ACTIVE_BORDER_TOP = registerColor('tab.unfocusedActiveBorderTop', {
    dark: ( transparent(TAB_ACTIVE_BORDER_TOP, 0.5)),
    light: ( transparent(TAB_ACTIVE_BORDER_TOP, 0.7)),
    hcDark: null,
    hcLight: '#B5200D'
}, ( localize(
    3810,
    "Border to the top of an active tab in an unfocused group. Tabs are the containers for editors in the editor area. Multiple tabs can be opened in one editor group. There can be multiple editor groups."
)));
const TAB_SELECTED_BORDER_TOP = registerColor('tab.selectedBorderTop', TAB_ACTIVE_BORDER_TOP, ( localize(
    3811,
    "Border to the top of a selected tab. Tabs are the containers for editors in the editor area. Multiple tabs can be opened in one editor group. There can be multiple editor groups."
)));
registerColor('tab.selectedBackground', TAB_ACTIVE_BACKGROUND, ( localize(
    3812,
    "Background of a selected tab. Tabs are the containers for editors in the editor area. Multiple tabs can be opened in one editor group. There can be multiple editor groups."
)));
registerColor('tab.selectedForeground', TAB_ACTIVE_FOREGROUND, ( localize(
    3813,
    "Foreground of a selected tab. Tabs are the containers for editors in the editor area. Multiple tabs can be opened in one editor group. There can be multiple editor groups."
)));
const TAB_HOVER_BORDER = registerColor('tab.hoverBorder', null, ( localize(
    3814,
    "Border to highlight tabs when hovering. Tabs are the containers for editors in the editor area. Multiple tabs can be opened in one editor group. There can be multiple editor groups."
)));
const TAB_UNFOCUSED_HOVER_BORDER = registerColor('tab.unfocusedHoverBorder', {
    dark: ( transparent(TAB_HOVER_BORDER, 0.5)),
    light: ( transparent(TAB_HOVER_BORDER, 0.7)),
    hcDark: null,
    hcLight: contrastBorder
}, ( localize(
    3815,
    "Border to highlight tabs in an unfocused group when hovering. Tabs are the containers for editors in the editor area. Multiple tabs can be opened in one editor group. There can be multiple editor groups."
)));
registerColor('tab.dragAndDropBorder', {
    dark: TAB_ACTIVE_FOREGROUND,
    light: TAB_ACTIVE_FOREGROUND,
    hcDark: activeContrastBorder,
    hcLight: activeContrastBorder
}, ( localize(
    3816,
    "Border between tabs to indicate that a tab can be inserted between two tabs. Tabs are the containers for editors in the editor area. Multiple tabs can be opened in one editor group. There can be multiple editor groups."
)));
const TAB_ACTIVE_MODIFIED_BORDER = registerColor('tab.activeModifiedBorder', {
    dark: '#3399CC',
    light: '#33AAEE',
    hcDark: null,
    hcLight: contrastBorder
}, ( localize(
    3817,
    "Border on the top of modified active tabs in an active group. Tabs are the containers for editors in the editor area. Multiple tabs can be opened in one editor group. There can be multiple editor groups."
)));
const TAB_INACTIVE_MODIFIED_BORDER = registerColor('tab.inactiveModifiedBorder', {
    dark: ( transparent(TAB_ACTIVE_MODIFIED_BORDER, 0.5)),
    light: ( transparent(TAB_ACTIVE_MODIFIED_BORDER, 0.5)),
    hcDark: Color.white,
    hcLight: contrastBorder
}, ( localize(
    3818,
    "Border on the top of modified inactive tabs in an active group. Tabs are the containers for editors in the editor area. Multiple tabs can be opened in one editor group. There can be multiple editor groups."
)));
const TAB_UNFOCUSED_ACTIVE_MODIFIED_BORDER = registerColor('tab.unfocusedActiveModifiedBorder', {
    dark: ( transparent(TAB_ACTIVE_MODIFIED_BORDER, 0.5)),
    light: ( transparent(TAB_ACTIVE_MODIFIED_BORDER, 0.7)),
    hcDark: Color.white,
    hcLight: contrastBorder
}, ( localize(
    3819,
    "Border on the top of modified active tabs in an unfocused group. Tabs are the containers for editors in the editor area. Multiple tabs can be opened in one editor group. There can be multiple editor groups."
)));
const TAB_UNFOCUSED_INACTIVE_MODIFIED_BORDER = registerColor('tab.unfocusedInactiveModifiedBorder', {
    dark: ( transparent(TAB_INACTIVE_MODIFIED_BORDER, 0.5)),
    light: ( transparent(TAB_INACTIVE_MODIFIED_BORDER, 0.5)),
    hcDark: Color.white,
    hcLight: contrastBorder
}, ( localize(
    3820,
    "Border on the top of modified inactive tabs in an unfocused group. Tabs are the containers for editors in the editor area. Multiple tabs can be opened in one editor group. There can be multiple editor groups."
)));
const EDITOR_PANE_BACKGROUND = registerColor('editorPane.background', editorBackground, ( localize(
    3821,
    "Background color of the editor pane visible on the left and right side of the centered editor layout."
)));
const EDITOR_GROUP_EMPTY_BACKGROUND = registerColor('editorGroup.emptyBackground', null, ( localize(
    3822,
    "Background color of an empty editor group. Editor groups are the containers of editors."
)));
registerColor('editorGroup.focusedEmptyBorder', {
    dark: null,
    light: null,
    hcDark: focusBorder,
    hcLight: focusBorder
}, ( localize(
    3823,
    "Border color of an empty editor group that is focused. Editor groups are the containers of editors."
)));
const EDITOR_GROUP_HEADER_TABS_BACKGROUND = registerColor('editorGroupHeader.tabsBackground', {
    dark: '#252526',
    light: '#F3F3F3',
    hcDark: null,
    hcLight: null
}, ( localize(
    3824,
    "Background color of the editor group title header when tabs are enabled. Editor groups are the containers of editors."
)));
const EDITOR_GROUP_HEADER_TABS_BORDER = registerColor('editorGroupHeader.tabsBorder', null, ( localize(
    3825,
    "Border color of the editor group title header when tabs are enabled. Editor groups are the containers of editors."
)));
const EDITOR_GROUP_HEADER_NO_TABS_BACKGROUND = registerColor('editorGroupHeader.noTabsBackground', editorBackground, ( localize(
    3826,
    "Background color of the editor group title header when (`\"workbench.editor.showTabs\": \"single\"`). Editor groups are the containers of editors."
)));
const EDITOR_GROUP_HEADER_BORDER = registerColor('editorGroupHeader.border', {
    dark: null,
    light: null,
    hcDark: contrastBorder,
    hcLight: contrastBorder
}, ( localize(
    3827,
    "Border color of the editor group title header. Editor groups are the containers of editors."
)));
const EDITOR_GROUP_BORDER = registerColor('editorGroup.border', {
    dark: '#444444',
    light: '#E7E7E7',
    hcDark: contrastBorder,
    hcLight: contrastBorder
}, ( localize(
    3828,
    "Color to separate multiple editor groups from each other. Editor groups are the containers of editors."
)));
const EDITOR_DRAG_AND_DROP_BACKGROUND = registerColor('editorGroup.dropBackground', {
    dark: ( ( Color.fromHex('#53595D')).transparent(0.5)),
    light: ( ( Color.fromHex('#2677CB')).transparent(0.18)),
    hcDark: null,
    hcLight: ( ( Color.fromHex('#0F4A85')).transparent(0.50))
}, ( localize(
    3829,
    "Background color when dragging editors around. The color should have transparency so that the editor contents can still shine through."
)));
const EDITOR_DROP_INTO_PROMPT_FOREGROUND = registerColor('editorGroup.dropIntoPromptForeground', editorWidgetForeground, ( localize(
    3830,
    "Foreground color of text shown over editors when dragging files. This text informs the user that they can hold shift to drop into the editor."
)));
const EDITOR_DROP_INTO_PROMPT_BACKGROUND = registerColor('editorGroup.dropIntoPromptBackground', editorWidgetBackground, ( localize(
    3831,
    "Background color of text shown over editors when dragging files. This text informs the user that they can hold shift to drop into the editor."
)));
const EDITOR_DROP_INTO_PROMPT_BORDER = registerColor('editorGroup.dropIntoPromptBorder', {
    dark: null,
    light: null,
    hcDark: contrastBorder,
    hcLight: contrastBorder
}, ( localize(
    3832,
    "Border color of text shown over editors when dragging files. This text informs the user that they can hold shift to drop into the editor."
)));
const SIDE_BY_SIDE_EDITOR_HORIZONTAL_BORDER = registerColor('sideBySideEditor.horizontalBorder', EDITOR_GROUP_BORDER, ( localize(
    3833,
    "Color to separate two editors from each other when shown side by side in an editor group from top to bottom."
)));
const SIDE_BY_SIDE_EDITOR_VERTICAL_BORDER = registerColor('sideBySideEditor.verticalBorder', EDITOR_GROUP_BORDER, ( localize(
    3834,
    "Color to separate two editors from each other when shown side by side in an editor group from left to right."
)));
const OUTPUT_VIEW_BACKGROUND = registerColor('outputView.background', null, ( localize(3835, "Output view background color.")));
registerColor('outputViewStickyScroll.background', OUTPUT_VIEW_BACKGROUND, ( localize(3836, "Output view sticky scroll background color.")));
registerColor('banner.background', {
    dark: listActiveSelectionBackground,
    light: ( darken(listActiveSelectionBackground, 0.3)),
    hcDark: listActiveSelectionBackground,
    hcLight: listActiveSelectionBackground
}, ( localize(
    3837,
    "Banner background color. The banner is shown under the title bar of the window."
)));
registerColor('banner.foreground', listActiveSelectionForeground, ( localize(
    3838,
    "Banner foreground color. The banner is shown under the title bar of the window."
)));
registerColor('banner.iconForeground', editorInfoForeground, ( localize(
    3839,
    "Banner icon color. The banner is shown under the title bar of the window."
)));
const STATUS_BAR_FOREGROUND = registerColor('statusBar.foreground', {
    dark: '#FFFFFF',
    light: '#FFFFFF',
    hcDark: '#FFFFFF',
    hcLight: editorForeground
}, ( localize(
    3840,
    "Status bar foreground color when a workspace or folder is opened. The status bar is shown in the bottom of the window."
)));
const STATUS_BAR_NO_FOLDER_FOREGROUND = registerColor('statusBar.noFolderForeground', STATUS_BAR_FOREGROUND, ( localize(
    3841,
    "Status bar foreground color when no folder is opened. The status bar is shown in the bottom of the window."
)));
const STATUS_BAR_BACKGROUND = registerColor('statusBar.background', {
    dark: '#007ACC',
    light: '#007ACC',
    hcDark: null,
    hcLight: null,
}, ( localize(
    3842,
    "Status bar background color when a workspace or folder is opened. The status bar is shown in the bottom of the window."
)));
const STATUS_BAR_NO_FOLDER_BACKGROUND = registerColor('statusBar.noFolderBackground', {
    dark: '#68217A',
    light: '#68217A',
    hcDark: null,
    hcLight: null,
}, ( localize(
    3843,
    "Status bar background color when no folder is opened. The status bar is shown in the bottom of the window."
)));
const STATUS_BAR_BORDER = registerColor('statusBar.border', {
    dark: null,
    light: null,
    hcDark: contrastBorder,
    hcLight: contrastBorder
}, ( localize(
    3844,
    "Status bar border color separating to the sidebar and editor. The status bar is shown in the bottom of the window."
)));
const STATUS_BAR_FOCUS_BORDER = registerColor('statusBar.focusBorder', {
    dark: STATUS_BAR_FOREGROUND,
    light: STATUS_BAR_FOREGROUND,
    hcDark: null,
    hcLight: STATUS_BAR_FOREGROUND
}, ( localize(
    3845,
    "Status bar border color when focused on keyboard navigation. The status bar is shown in the bottom of the window."
)));
const STATUS_BAR_NO_FOLDER_BORDER = registerColor('statusBar.noFolderBorder', STATUS_BAR_BORDER, ( localize(
    3846,
    "Status bar border color separating to the sidebar and editor when no folder is opened. The status bar is shown in the bottom of the window."
)));
registerColor('statusBarItem.activeBackground', {
    dark: ( Color.white.transparent(0.18)),
    light: ( Color.white.transparent(0.18)),
    hcDark: ( Color.white.transparent(0.18)),
    hcLight: ( Color.black.transparent(0.18))
}, ( localize(
    3847,
    "Status bar item background color when clicking. The status bar is shown in the bottom of the window."
)));
const STATUS_BAR_ITEM_FOCUS_BORDER = registerColor('statusBarItem.focusBorder', {
    dark: STATUS_BAR_FOREGROUND,
    light: STATUS_BAR_FOREGROUND,
    hcDark: null,
    hcLight: activeContrastBorder
}, ( localize(
    3848,
    "Status bar item border color when focused on keyboard navigation. The status bar is shown in the bottom of the window."
)));
const STATUS_BAR_ITEM_HOVER_BACKGROUND = registerColor('statusBarItem.hoverBackground', {
    dark: ( Color.white.transparent(0.12)),
    light: ( Color.white.transparent(0.12)),
    hcDark: ( Color.white.transparent(0.12)),
    hcLight: ( Color.black.transparent(0.12))
}, ( localize(
    3849,
    "Status bar item background color when hovering. The status bar is shown in the bottom of the window."
)));
const STATUS_BAR_ITEM_HOVER_FOREGROUND = registerColor('statusBarItem.hoverForeground', STATUS_BAR_FOREGROUND, ( localize(
    3850,
    "Status bar item foreground color when hovering. The status bar is shown in the bottom of the window."
)));
const STATUS_BAR_ITEM_COMPACT_HOVER_BACKGROUND = registerColor('statusBarItem.compactHoverBackground', {
    dark: ( Color.white.transparent(0.20)),
    light: ( Color.white.transparent(0.20)),
    hcDark: ( Color.white.transparent(0.20)),
    hcLight: ( Color.black.transparent(0.20))
}, ( localize(
    3851,
    "Status bar item background color when hovering an item that contains two hovers. The status bar is shown in the bottom of the window."
)));
registerColor('statusBarItem.prominentForeground', STATUS_BAR_FOREGROUND, ( localize(
    3852,
    "Status bar prominent items foreground color. Prominent items stand out from other status bar entries to indicate importance. The status bar is shown in the bottom of the window."
)));
registerColor('statusBarItem.prominentBackground', ( Color.black.transparent(0.5)), ( localize(
    3853,
    "Status bar prominent items background color. Prominent items stand out from other status bar entries to indicate importance. The status bar is shown in the bottom of the window."
)));
registerColor('statusBarItem.prominentHoverForeground', STATUS_BAR_ITEM_HOVER_FOREGROUND, ( localize(
    3854,
    "Status bar prominent items foreground color when hovering. Prominent items stand out from other status bar entries to indicate importance. The status bar is shown in the bottom of the window."
)));
registerColor('statusBarItem.prominentHoverBackground', STATUS_BAR_ITEM_HOVER_BACKGROUND, ( localize(
    3855,
    "Status bar prominent items background color when hovering. Prominent items stand out from other status bar entries to indicate importance. The status bar is shown in the bottom of the window."
)));
const STATUS_BAR_ERROR_ITEM_BACKGROUND = registerColor('statusBarItem.errorBackground', {
    dark: ( darken(errorForeground, .4)),
    light: ( darken(errorForeground, .4)),
    hcDark: null,
    hcLight: '#B5200D'
}, ( localize(
    3856,
    "Status bar error items background color. Error items stand out from other status bar entries to indicate error conditions. The status bar is shown in the bottom of the window."
)));
registerColor('statusBarItem.errorForeground', Color.white, ( localize(
    3857,
    "Status bar error items foreground color. Error items stand out from other status bar entries to indicate error conditions. The status bar is shown in the bottom of the window."
)));
registerColor('statusBarItem.errorHoverForeground', STATUS_BAR_ITEM_HOVER_FOREGROUND, ( localize(
    3858,
    "Status bar error items foreground color when hovering. Error items stand out from other status bar entries to indicate error conditions. The status bar is shown in the bottom of the window."
)));
registerColor('statusBarItem.errorHoverBackground', STATUS_BAR_ITEM_HOVER_BACKGROUND, ( localize(
    3859,
    "Status bar error items background color when hovering. Error items stand out from other status bar entries to indicate error conditions. The status bar is shown in the bottom of the window."
)));
const STATUS_BAR_WARNING_ITEM_BACKGROUND = registerColor('statusBarItem.warningBackground', {
    dark: ( darken(editorWarningForeground, .4)),
    light: ( darken(editorWarningForeground, .4)),
    hcDark: null,
    hcLight: '#895503'
}, ( localize(
    3860,
    "Status bar warning items background color. Warning items stand out from other status bar entries to indicate warning conditions. The status bar is shown in the bottom of the window."
)));
registerColor('statusBarItem.warningForeground', Color.white, ( localize(
    3861,
    "Status bar warning items foreground color. Warning items stand out from other status bar entries to indicate warning conditions. The status bar is shown in the bottom of the window."
)));
registerColor('statusBarItem.warningHoverForeground', STATUS_BAR_ITEM_HOVER_FOREGROUND, ( localize(
    3862,
    "Status bar warning items foreground color when hovering. Warning items stand out from other status bar entries to indicate warning conditions. The status bar is shown in the bottom of the window."
)));
registerColor('statusBarItem.warningHoverBackground', STATUS_BAR_ITEM_HOVER_BACKGROUND, ( localize(
    3863,
    "Status bar warning items background color when hovering. Warning items stand out from other status bar entries to indicate warning conditions. The status bar is shown in the bottom of the window."
)));
const ACTIVITY_BAR_BACKGROUND = registerColor('activityBar.background', {
    dark: '#333333',
    light: '#2C2C2C',
    hcDark: '#000000',
    hcLight: '#FFFFFF'
}, ( localize(
    3864,
    "Activity bar background color. The activity bar is showing on the far left or right and allows to switch between views of the side bar."
)));
const ACTIVITY_BAR_FOREGROUND = registerColor('activityBar.foreground', {
    dark: Color.white,
    light: Color.white,
    hcDark: Color.white,
    hcLight: editorForeground
}, ( localize(
    3865,
    "Activity bar item foreground color when it is active. The activity bar is showing on the far left or right and allows to switch between views of the side bar."
)));
const ACTIVITY_BAR_INACTIVE_FOREGROUND = registerColor('activityBar.inactiveForeground', {
    dark: ( transparent(ACTIVITY_BAR_FOREGROUND, 0.4)),
    light: ( transparent(ACTIVITY_BAR_FOREGROUND, 0.4)),
    hcDark: Color.white,
    hcLight: editorForeground
}, ( localize(
    3866,
    "Activity bar item foreground color when it is inactive. The activity bar is showing on the far left or right and allows to switch between views of the side bar."
)));
const ACTIVITY_BAR_BORDER = registerColor('activityBar.border', {
    dark: null,
    light: null,
    hcDark: contrastBorder,
    hcLight: contrastBorder
}, ( localize(
    3867,
    "Activity bar border color separating to the side bar. The activity bar is showing on the far left or right and allows to switch between views of the side bar."
)));
const ACTIVITY_BAR_ACTIVE_BORDER = registerColor('activityBar.activeBorder', {
    dark: ACTIVITY_BAR_FOREGROUND,
    light: ACTIVITY_BAR_FOREGROUND,
    hcDark: contrastBorder,
    hcLight: contrastBorder
}, ( localize(
    3868,
    "Activity bar border color for the active item. The activity bar is showing on the far left or right and allows to switch between views of the side bar."
)));
const ACTIVITY_BAR_ACTIVE_FOCUS_BORDER = registerColor('activityBar.activeFocusBorder', {
    dark: null,
    light: null,
    hcDark: null,
    hcLight: '#B5200D'
}, ( localize(
    3869,
    "Activity bar focus border color for the active item. The activity bar is showing on the far left or right and allows to switch between views of the side bar."
)));
const ACTIVITY_BAR_ACTIVE_BACKGROUND = registerColor('activityBar.activeBackground', null, ( localize(
    3870,
    "Activity bar background color for the active item. The activity bar is showing on the far left or right and allows to switch between views of the side bar."
)));
const ACTIVITY_BAR_DRAG_AND_DROP_BORDER = registerColor('activityBar.dropBorder', {
    dark: ACTIVITY_BAR_FOREGROUND,
    light: ACTIVITY_BAR_FOREGROUND,
    hcDark: null,
    hcLight: null,
}, ( localize(
    3871,
    "Drag and drop feedback color for the activity bar items. The activity bar is showing on the far left or right and allows to switch between views of the side bar."
)));
const ACTIVITY_BAR_BADGE_BACKGROUND = registerColor('activityBarBadge.background', {
    dark: '#007ACC',
    light: '#007ACC',
    hcDark: '#000000',
    hcLight: '#0F4A85'
}, ( localize(
    3872,
    "Activity notification badge background color. The activity bar is showing on the far left or right and allows to switch between views of the side bar."
)));
const ACTIVITY_BAR_BADGE_FOREGROUND = registerColor('activityBarBadge.foreground', Color.white, ( localize(
    3873,
    "Activity notification badge foreground color. The activity bar is showing on the far left or right and allows to switch between views of the side bar."
)));
const ACTIVITY_BAR_TOP_FOREGROUND = registerColor('activityBarTop.foreground', {
    dark: '#E7E7E7',
    light: '#424242',
    hcDark: Color.white,
    hcLight: editorForeground
}, ( localize(
    3874,
    "Active foreground color of the item in the Activity bar when it is on top / bottom. The activity allows to switch between views of the side bar."
)));
const ACTIVITY_BAR_TOP_ACTIVE_BORDER = registerColor('activityBarTop.activeBorder', {
    dark: ACTIVITY_BAR_TOP_FOREGROUND,
    light: ACTIVITY_BAR_TOP_FOREGROUND,
    hcDark: contrastBorder,
    hcLight: '#B5200D'
}, ( localize(
    3875,
    "Focus border color for the active item in the Activity bar when it is on top / bottom. The activity allows to switch between views of the side bar."
)));
registerColor('activityBarTop.activeBackground', null, ( localize(
    3876,
    "Background color for the active item in the Activity bar when it is on top / bottom. The activity allows to switch between views of the side bar."
)));
const ACTIVITY_BAR_TOP_INACTIVE_FOREGROUND = registerColor('activityBarTop.inactiveForeground', {
    dark: ( transparent(ACTIVITY_BAR_TOP_FOREGROUND, 0.6)),
    light: ( transparent(ACTIVITY_BAR_TOP_FOREGROUND, 0.75)),
    hcDark: Color.white,
    hcLight: editorForeground
}, ( localize(
    3877,
    "Inactive foreground color of the item in the Activity bar when it is on top / bottom. The activity allows to switch between views of the side bar."
)));
const ACTIVITY_BAR_TOP_DRAG_AND_DROP_BORDER = registerColor('activityBarTop.dropBorder', ACTIVITY_BAR_TOP_FOREGROUND, ( localize(
    3878,
    "Drag and drop feedback color for the items in the Activity bar when it is on top / bottom. The activity allows to switch between views of the side bar."
)));
registerColor('activityBarTop.background', null, ( localize(3879, "Background color of the activity bar when set to top / bottom.")));
const PANEL_BACKGROUND = registerColor('panel.background', editorBackground, ( localize(
    3880,
    "Panel background color. Panels are shown below the editor area and contain views like output and integrated terminal."
)));
const PANEL_BORDER = registerColor('panel.border', {
    dark: ( ( Color.fromHex('#808080')).transparent(0.35)),
    light: ( ( Color.fromHex('#808080')).transparent(0.35)),
    hcDark: contrastBorder,
    hcLight: contrastBorder
}, ( localize(
    3881,
    "Panel border color to separate the panel from the editor. Panels are shown below the editor area and contain views like output and integrated terminal."
)));
const PANEL_TITLE_BORDER = registerColor('panelTitle.border', {
    dark: null,
    light: null,
    hcDark: PANEL_BORDER,
    hcLight: PANEL_BORDER
}, ( localize(
    3882,
    "Panel title border color on the bottom, separating the title from the views. Panels are shown below the editor area and contain views like output and integrated terminal."
)));
const PANEL_ACTIVE_TITLE_FOREGROUND = registerColor('panelTitle.activeForeground', {
    dark: '#E7E7E7',
    light: '#424242',
    hcDark: Color.white,
    hcLight: editorForeground
}, ( localize(
    3883,
    "Title color for the active panel. Panels are shown below the editor area and contain views like output and integrated terminal."
)));
const PANEL_INACTIVE_TITLE_FOREGROUND = registerColor('panelTitle.inactiveForeground', {
    dark: ( transparent(PANEL_ACTIVE_TITLE_FOREGROUND, 0.6)),
    light: ( transparent(PANEL_ACTIVE_TITLE_FOREGROUND, 0.75)),
    hcDark: Color.white,
    hcLight: editorForeground
}, ( localize(
    3884,
    "Title color for the inactive panel. Panels are shown below the editor area and contain views like output and integrated terminal."
)));
const PANEL_ACTIVE_TITLE_BORDER = registerColor('panelTitle.activeBorder', {
    dark: PANEL_ACTIVE_TITLE_FOREGROUND,
    light: PANEL_ACTIVE_TITLE_FOREGROUND,
    hcDark: contrastBorder,
    hcLight: '#B5200D'
}, ( localize(
    3885,
    "Border color for the active panel title. Panels are shown below the editor area and contain views like output and integrated terminal."
)));
const PANEL_TITLE_BADGE_BACKGROUND = registerColor('panelTitleBadge.background', ACTIVITY_BAR_BADGE_BACKGROUND, ( localize(
    3886,
    "Panel title badge background color. Panels are shown below the editor area and contain views like output and integrated terminal."
)));
const PANEL_TITLE_BADGE_FOREGROUND = registerColor('panelTitleBadge.foreground', ACTIVITY_BAR_BADGE_FOREGROUND, ( localize(
    3887,
    "Panel title badge foreground color. Panels are shown below the editor area and contain views like output and integrated terminal."
)));
registerColor('panelInput.border', {
    dark: inputBorder,
    light: ( Color.fromHex('#ddd')),
    hcDark: inputBorder,
    hcLight: inputBorder
}, ( localize(3888, "Input box border for inputs in the panel.")));
const PANEL_DRAG_AND_DROP_BORDER = registerColor('panel.dropBorder', PANEL_ACTIVE_TITLE_FOREGROUND, ( localize(
    3889,
    "Drag and drop feedback color for the panel titles. Panels are shown below the editor area and contain views like output and integrated terminal."
)));
const PANEL_SECTION_DRAG_AND_DROP_BACKGROUND = registerColor('panelSection.dropBackground', EDITOR_DRAG_AND_DROP_BACKGROUND, ( localize(
    3890,
    "Drag and drop feedback color for the panel sections. The color should have transparency so that the panel sections can still shine through. Panels are shown below the editor area and contain views like output and integrated terminal. Panel sections are views nested within the panels."
)));
const PANEL_SECTION_HEADER_BACKGROUND = registerColor('panelSectionHeader.background', {
    dark: ( ( Color.fromHex('#808080')).transparent(0.2)),
    light: ( ( Color.fromHex('#808080')).transparent(0.2)),
    hcDark: null,
    hcLight: null,
}, ( localize(
    3891,
    "Panel section header background color. Panels are shown below the editor area and contain views like output and integrated terminal. Panel sections are views nested within the panels."
)));
const PANEL_SECTION_HEADER_FOREGROUND = registerColor('panelSectionHeader.foreground', null, ( localize(
    3892,
    "Panel section header foreground color. Panels are shown below the editor area and contain views like output and integrated terminal. Panel sections are views nested within the panels."
)));
const PANEL_SECTION_HEADER_BORDER = registerColor('panelSectionHeader.border', contrastBorder, ( localize(
    3893,
    "Panel section header border color used when multiple views are stacked vertically in the panel. Panels are shown below the editor area and contain views like output and integrated terminal. Panel sections are views nested within the panels."
)));
const PANEL_SECTION_BORDER = registerColor('panelSection.border', PANEL_BORDER, ( localize(
    3894,
    "Panel section border color used when multiple views are stacked horizontally in the panel. Panels are shown below the editor area and contain views like output and integrated terminal. Panel sections are views nested within the panels."
)));
const PANEL_STICKY_SCROLL_BACKGROUND = registerColor('panelStickyScroll.background', PANEL_BACKGROUND, ( localize(3895, "Background color of sticky scroll in the panel.")));
const PANEL_STICKY_SCROLL_BORDER = registerColor('panelStickyScroll.border', null, ( localize(3896, "Border color of sticky scroll in the panel.")));
const PANEL_STICKY_SCROLL_SHADOW = registerColor('panelStickyScroll.shadow', scrollbarShadow, ( localize(3897, "Shadow color of sticky scroll in the panel.")));
registerColor('profileBadge.background', {
    dark: '#4D4D4D',
    light: '#C4C4C4',
    hcDark: Color.white,
    hcLight: Color.black
}, ( localize(
    3898,
    "Profile badge background color. The profile badge shows on top of the settings gear icon in the activity bar."
)));
registerColor('profileBadge.foreground', {
    dark: Color.white,
    light: '#333333',
    hcDark: Color.black,
    hcLight: Color.white
}, ( localize(
    3899,
    "Profile badge foreground color. The profile badge shows on top of the settings gear icon in the activity bar."
)));
const STATUS_BAR_REMOTE_ITEM_BACKGROUND = registerColor('statusBarItem.remoteBackground', ACTIVITY_BAR_BADGE_BACKGROUND, ( localize(3900, "Background color for the remote indicator on the status bar.")));
const STATUS_BAR_REMOTE_ITEM_FOREGROUND = registerColor('statusBarItem.remoteForeground', ACTIVITY_BAR_BADGE_FOREGROUND, ( localize(3901, "Foreground color for the remote indicator on the status bar.")));
registerColor('statusBarItem.remoteHoverForeground', STATUS_BAR_ITEM_HOVER_FOREGROUND, ( localize(
    3902,
    "Foreground color for the remote indicator on the status bar when hovering."
)));
registerColor('statusBarItem.remoteHoverBackground', {
    dark: STATUS_BAR_ITEM_HOVER_BACKGROUND,
    light: STATUS_BAR_ITEM_HOVER_BACKGROUND,
    hcDark: STATUS_BAR_ITEM_HOVER_BACKGROUND,
    hcLight: null
}, ( localize(
    3903,
    "Background color for the remote indicator on the status bar when hovering."
)));
registerColor('statusBarItem.offlineBackground', '#6c1717', ( localize(3904, "Status bar item background color when the workbench is offline.")));
registerColor('statusBarItem.offlineForeground', STATUS_BAR_REMOTE_ITEM_FOREGROUND, ( localize(3905, "Status bar item foreground color when the workbench is offline.")));
registerColor('statusBarItem.offlineHoverForeground', STATUS_BAR_ITEM_HOVER_FOREGROUND, ( localize(
    3906,
    "Status bar item foreground hover color when the workbench is offline."
)));
registerColor('statusBarItem.offlineHoverBackground', {
    dark: STATUS_BAR_ITEM_HOVER_BACKGROUND,
    light: STATUS_BAR_ITEM_HOVER_BACKGROUND,
    hcDark: STATUS_BAR_ITEM_HOVER_BACKGROUND,
    hcLight: null
}, ( localize(
    3907,
    "Status bar item background hover color when the workbench is offline."
)));
const EXTENSION_BADGE_REMOTE_BACKGROUND = registerColor('extensionBadge.remoteBackground', ACTIVITY_BAR_BADGE_BACKGROUND, ( localize(3908, "Background color for the remote badge in the extensions view.")));
const EXTENSION_BADGE_REMOTE_FOREGROUND = registerColor('extensionBadge.remoteForeground', ACTIVITY_BAR_BADGE_FOREGROUND, ( localize(3909, "Foreground color for the remote badge in the extensions view.")));
const SIDE_BAR_BACKGROUND = registerColor('sideBar.background', {
    dark: '#252526',
    light: '#F3F3F3',
    hcDark: '#000000',
    hcLight: '#FFFFFF'
}, ( localize(
    3910,
    "Side bar background color. The side bar is the container for views like explorer and search."
)));
const SIDE_BAR_FOREGROUND = registerColor('sideBar.foreground', null, ( localize(
    3911,
    "Side bar foreground color. The side bar is the container for views like explorer and search."
)));
const SIDE_BAR_BORDER = registerColor('sideBar.border', {
    dark: null,
    light: null,
    hcDark: contrastBorder,
    hcLight: contrastBorder
}, ( localize(
    3912,
    "Side bar border color on the side separating to the editor. The side bar is the container for views like explorer and search."
)));
registerColor('sideBarTitle.background', SIDE_BAR_BACKGROUND, ( localize(
    3913,
    "Side bar title background color. The side bar is the container for views like explorer and search."
)));
const SIDE_BAR_TITLE_FOREGROUND = registerColor('sideBarTitle.foreground', SIDE_BAR_FOREGROUND, ( localize(
    3914,
    "Side bar title foreground color. The side bar is the container for views like explorer and search."
)));
const SIDE_BAR_TITLE_BORDER = registerColor('sideBarTitle.border', {
    dark: null,
    light: null,
    hcDark: SIDE_BAR_BORDER,
    hcLight: SIDE_BAR_BORDER
}, ( localize(
    3915,
    "Side bar title border color on the bottom, separating the title from the views. The side bar is the container for views like explorer and search."
)));
const SIDE_BAR_DRAG_AND_DROP_BACKGROUND = registerColor('sideBar.dropBackground', EDITOR_DRAG_AND_DROP_BACKGROUND, ( localize(
    3916,
    "Drag and drop feedback color for the side bar sections. The color should have transparency so that the side bar sections can still shine through. The side bar is the container for views like explorer and search. Side bar sections are views nested within the side bar."
)));
const SIDE_BAR_SECTION_HEADER_BACKGROUND = registerColor('sideBarSectionHeader.background', {
    dark: ( ( Color.fromHex('#808080')).transparent(0.2)),
    light: ( ( Color.fromHex('#808080')).transparent(0.2)),
    hcDark: null,
    hcLight: null
}, ( localize(
    3917,
    "Side bar section header background color. The side bar is the container for views like explorer and search. Side bar sections are views nested within the side bar."
)));
const SIDE_BAR_SECTION_HEADER_FOREGROUND = registerColor('sideBarSectionHeader.foreground', SIDE_BAR_FOREGROUND, ( localize(
    3918,
    "Side bar section header foreground color. The side bar is the container for views like explorer and search. Side bar sections are views nested within the side bar."
)));
const SIDE_BAR_SECTION_HEADER_BORDER = registerColor('sideBarSectionHeader.border', contrastBorder, ( localize(
    3919,
    "Side bar section header border color. The side bar is the container for views like explorer and search. Side bar sections are views nested within the side bar."
)));
registerColor('sideBarActivityBarTop.border', SIDE_BAR_SECTION_HEADER_BORDER, ( localize(
    3920,
    "Border color between the activity bar at the top/bottom and the views."
)));
const SIDE_BAR_STICKY_SCROLL_BACKGROUND = registerColor('sideBarStickyScroll.background', SIDE_BAR_BACKGROUND, ( localize(3921, "Background color of sticky scroll in the side bar.")));
const SIDE_BAR_STICKY_SCROLL_BORDER = registerColor('sideBarStickyScroll.border', null, ( localize(3922, "Border color of sticky scroll in the side bar.")));
const SIDE_BAR_STICKY_SCROLL_SHADOW = registerColor('sideBarStickyScroll.shadow', scrollbarShadow, ( localize(3923, "Shadow color of sticky scroll in the side bar.")));
const TITLE_BAR_ACTIVE_FOREGROUND = registerColor('titleBar.activeForeground', {
    dark: '#CCCCCC',
    light: '#333333',
    hcDark: '#FFFFFF',
    hcLight: '#292929'
}, ( localize(3924, "Title bar foreground when the window is active.")));
const TITLE_BAR_INACTIVE_FOREGROUND = registerColor('titleBar.inactiveForeground', {
    dark: ( transparent(TITLE_BAR_ACTIVE_FOREGROUND, 0.6)),
    light: ( transparent(TITLE_BAR_ACTIVE_FOREGROUND, 0.6)),
    hcDark: null,
    hcLight: '#292929'
}, ( localize(3925, "Title bar foreground when the window is inactive.")));
const TITLE_BAR_ACTIVE_BACKGROUND = registerColor('titleBar.activeBackground', {
    dark: '#3C3C3C',
    light: '#DDDDDD',
    hcDark: '#000000',
    hcLight: '#FFFFFF'
}, ( localize(3926, "Title bar background when the window is active.")));
const TITLE_BAR_INACTIVE_BACKGROUND = registerColor('titleBar.inactiveBackground', {
    dark: ( transparent(TITLE_BAR_ACTIVE_BACKGROUND, 0.6)),
    light: ( transparent(TITLE_BAR_ACTIVE_BACKGROUND, 0.6)),
    hcDark: null,
    hcLight: null,
}, ( localize(3927, "Title bar background when the window is inactive.")));
const TITLE_BAR_BORDER = registerColor('titleBar.border', {
    dark: null,
    light: null,
    hcDark: contrastBorder,
    hcLight: contrastBorder
}, ( localize(3928, "Title bar border color.")));
const MENUBAR_SELECTION_FOREGROUND = registerColor('menubar.selectionForeground', TITLE_BAR_ACTIVE_FOREGROUND, ( localize(3929, "Foreground color of the selected menu item in the menubar.")));
const MENUBAR_SELECTION_BACKGROUND = registerColor('menubar.selectionBackground', {
    dark: toolbarHoverBackground,
    light: toolbarHoverBackground,
    hcDark: null,
    hcLight: null,
}, ( localize(3930, "Background color of the selected menu item in the menubar.")));
registerColor('menubar.selectionBorder', {
    dark: null,
    light: null,
    hcDark: activeContrastBorder,
    hcLight: activeContrastBorder,
}, ( localize(3931, "Border color of the selected menu item in the menubar.")));
registerColor('commandCenter.foreground', TITLE_BAR_ACTIVE_FOREGROUND, ( localize(3932, "Foreground color of the command center")), false);
registerColor('commandCenter.activeForeground', MENUBAR_SELECTION_FOREGROUND, ( localize(3933, "Active foreground color of the command center")), false);
registerColor('commandCenter.inactiveForeground', TITLE_BAR_INACTIVE_FOREGROUND, ( localize(3934, "Foreground color of the command center when the window is inactive")), false);
const COMMAND_CENTER_BACKGROUND = registerColor('commandCenter.background', { dark: ( Color.white.transparent(0.05)), hcDark: null, light: ( Color.black.transparent(0.05)), hcLight: null }, ( localize(3935, "Background color of the command center")), false);
registerColor('commandCenter.activeBackground', { dark: ( Color.white.transparent(0.08)), hcDark: MENUBAR_SELECTION_BACKGROUND, light: ( Color.black.transparent(0.08)), hcLight: MENUBAR_SELECTION_BACKGROUND }, ( localize(3936, "Active background color of the command center")), false);
registerColor('commandCenter.border', { dark: ( transparent(TITLE_BAR_ACTIVE_FOREGROUND, .20)), hcDark: contrastBorder, light: ( transparent(TITLE_BAR_ACTIVE_FOREGROUND, .20)), hcLight: contrastBorder }, ( localize(3937, "Border color of the command center")), false);
registerColor('commandCenter.activeBorder', { dark: ( transparent(TITLE_BAR_ACTIVE_FOREGROUND, .30)), hcDark: TITLE_BAR_ACTIVE_FOREGROUND, light: ( transparent(TITLE_BAR_ACTIVE_FOREGROUND, .30)), hcLight: TITLE_BAR_ACTIVE_FOREGROUND }, ( localize(3938, "Active border color of the command center")), false);
registerColor('commandCenter.inactiveBorder', ( transparent(TITLE_BAR_INACTIVE_FOREGROUND, .25)), ( localize(3939, "Border color of the command center when the window is inactive")), false);
const NOTIFICATIONS_CENTER_BORDER = registerColor('notificationCenter.border', {
    dark: widgetBorder,
    light: widgetBorder,
    hcDark: contrastBorder,
    hcLight: contrastBorder
}, ( localize(
    3940,
    "Notifications center border color. Notifications slide in from the bottom right of the window."
)));
const NOTIFICATIONS_TOAST_BORDER = registerColor('notificationToast.border', {
    dark: widgetBorder,
    light: widgetBorder,
    hcDark: contrastBorder,
    hcLight: contrastBorder
}, ( localize(
    3941,
    "Notification toast border color. Notifications slide in from the bottom right of the window."
)));
registerColor('notifications.foreground', editorWidgetForeground, ( localize(
    3942,
    "Notifications foreground color. Notifications slide in from the bottom right of the window."
)));
const NOTIFICATIONS_BACKGROUND = registerColor('notifications.background', editorWidgetBackground, ( localize(
    3943,
    "Notifications background color. Notifications slide in from the bottom right of the window."
)));
registerColor('notificationLink.foreground', textLinkForeground, ( localize(
    3944,
    "Notification links foreground color. Notifications slide in from the bottom right of the window."
)));
const NOTIFICATIONS_CENTER_HEADER_FOREGROUND = registerColor('notificationCenterHeader.foreground', null, ( localize(
    3945,
    "Notifications center header foreground color. Notifications slide in from the bottom right of the window."
)));
const NOTIFICATIONS_CENTER_HEADER_BACKGROUND = registerColor('notificationCenterHeader.background', {
    dark: ( lighten(NOTIFICATIONS_BACKGROUND, 0.3)),
    light: ( darken(NOTIFICATIONS_BACKGROUND, 0.05)),
    hcDark: NOTIFICATIONS_BACKGROUND,
    hcLight: NOTIFICATIONS_BACKGROUND
}, ( localize(
    3946,
    "Notifications center header background color. Notifications slide in from the bottom right of the window."
)));
registerColor('notifications.border', NOTIFICATIONS_CENTER_HEADER_BACKGROUND, ( localize(
    3947,
    "Notifications border color separating from other notifications in the notifications center. Notifications slide in from the bottom right of the window."
)));
registerColor('notificationsErrorIcon.foreground', editorErrorForeground, ( localize(
    3948,
    "The color used for the icon of error notifications. Notifications slide in from the bottom right of the window."
)));
registerColor('notificationsWarningIcon.foreground', editorWarningForeground, ( localize(
    3949,
    "The color used for the icon of warning notifications. Notifications slide in from the bottom right of the window."
)));
registerColor('notificationsInfoIcon.foreground', editorInfoForeground, ( localize(
    3950,
    "The color used for the icon of info notifications. Notifications slide in from the bottom right of the window."
)));
const WINDOW_ACTIVE_BORDER = registerColor('window.activeBorder', {
    dark: null,
    light: null,
    hcDark: contrastBorder,
    hcLight: contrastBorder
}, ( localize(
    3951,
    "The color used for the border of the window when it is active. Only supported in the macOS and Linux desktop client when using the custom title bar."
)));
const WINDOW_INACTIVE_BORDER = registerColor('window.inactiveBorder', {
    dark: null,
    light: null,
    hcDark: contrastBorder,
    hcLight: contrastBorder
}, ( localize(
    3952,
    "The color used for the border of the window when it is inactive. Only supported in the macOS and Linux desktop client when using the custom title bar."
)));

export { ACTIVITY_BAR_ACTIVE_BACKGROUND, ACTIVITY_BAR_ACTIVE_BORDER, ACTIVITY_BAR_ACTIVE_FOCUS_BORDER, ACTIVITY_BAR_BACKGROUND, ACTIVITY_BAR_BADGE_BACKGROUND, ACTIVITY_BAR_BADGE_FOREGROUND, ACTIVITY_BAR_BORDER, ACTIVITY_BAR_DRAG_AND_DROP_BORDER, ACTIVITY_BAR_FOREGROUND, ACTIVITY_BAR_INACTIVE_FOREGROUND, ACTIVITY_BAR_TOP_ACTIVE_BORDER, ACTIVITY_BAR_TOP_DRAG_AND_DROP_BORDER, ACTIVITY_BAR_TOP_FOREGROUND, ACTIVITY_BAR_TOP_INACTIVE_FOREGROUND, COMMAND_CENTER_BACKGROUND, EDITOR_DRAG_AND_DROP_BACKGROUND, EDITOR_DROP_INTO_PROMPT_BACKGROUND, EDITOR_DROP_INTO_PROMPT_BORDER, EDITOR_DROP_INTO_PROMPT_FOREGROUND, EDITOR_GROUP_BORDER, EDITOR_GROUP_EMPTY_BACKGROUND, EDITOR_GROUP_HEADER_BORDER, EDITOR_GROUP_HEADER_NO_TABS_BACKGROUND, EDITOR_GROUP_HEADER_TABS_BACKGROUND, EDITOR_GROUP_HEADER_TABS_BORDER, EDITOR_PANE_BACKGROUND, EXTENSION_BADGE_REMOTE_BACKGROUND, EXTENSION_BADGE_REMOTE_FOREGROUND, MENUBAR_SELECTION_BACKGROUND, MENUBAR_SELECTION_FOREGROUND, NOTIFICATIONS_BACKGROUND, NOTIFICATIONS_CENTER_BORDER, NOTIFICATIONS_CENTER_HEADER_BACKGROUND, NOTIFICATIONS_CENTER_HEADER_FOREGROUND, NOTIFICATIONS_TOAST_BORDER, PANEL_ACTIVE_TITLE_BORDER, PANEL_ACTIVE_TITLE_FOREGROUND, PANEL_BACKGROUND, PANEL_BORDER, PANEL_DRAG_AND_DROP_BORDER, PANEL_INACTIVE_TITLE_FOREGROUND, PANEL_SECTION_BORDER, PANEL_SECTION_DRAG_AND_DROP_BACKGROUND, PANEL_SECTION_HEADER_BACKGROUND, PANEL_SECTION_HEADER_BORDER, PANEL_SECTION_HEADER_FOREGROUND, PANEL_STICKY_SCROLL_BACKGROUND, PANEL_STICKY_SCROLL_BORDER, PANEL_STICKY_SCROLL_SHADOW, PANEL_TITLE_BADGE_BACKGROUND, PANEL_TITLE_BADGE_FOREGROUND, PANEL_TITLE_BORDER, SIDE_BAR_BACKGROUND, SIDE_BAR_BORDER, SIDE_BAR_DRAG_AND_DROP_BACKGROUND, SIDE_BAR_FOREGROUND, SIDE_BAR_SECTION_HEADER_BACKGROUND, SIDE_BAR_SECTION_HEADER_BORDER, SIDE_BAR_SECTION_HEADER_FOREGROUND, SIDE_BAR_STICKY_SCROLL_BACKGROUND, SIDE_BAR_STICKY_SCROLL_BORDER, SIDE_BAR_STICKY_SCROLL_SHADOW, SIDE_BAR_TITLE_BORDER, SIDE_BAR_TITLE_FOREGROUND, SIDE_BY_SIDE_EDITOR_HORIZONTAL_BORDER, SIDE_BY_SIDE_EDITOR_VERTICAL_BORDER, STATUS_BAR_BACKGROUND, STATUS_BAR_BORDER, STATUS_BAR_ERROR_ITEM_BACKGROUND, STATUS_BAR_FOCUS_BORDER, STATUS_BAR_FOREGROUND, STATUS_BAR_ITEM_COMPACT_HOVER_BACKGROUND, STATUS_BAR_ITEM_FOCUS_BORDER, STATUS_BAR_ITEM_HOVER_BACKGROUND, STATUS_BAR_ITEM_HOVER_FOREGROUND, STATUS_BAR_NO_FOLDER_BACKGROUND, STATUS_BAR_NO_FOLDER_BORDER, STATUS_BAR_NO_FOLDER_FOREGROUND, STATUS_BAR_REMOTE_ITEM_BACKGROUND, STATUS_BAR_REMOTE_ITEM_FOREGROUND, STATUS_BAR_WARNING_ITEM_BACKGROUND, TAB_ACTIVE_BACKGROUND, TAB_ACTIVE_BORDER, TAB_ACTIVE_BORDER_TOP, TAB_ACTIVE_FOREGROUND, TAB_ACTIVE_MODIFIED_BORDER, TAB_BORDER, TAB_HOVER_BACKGROUND, TAB_HOVER_BORDER, TAB_HOVER_FOREGROUND, TAB_INACTIVE_BACKGROUND, TAB_INACTIVE_FOREGROUND, TAB_INACTIVE_MODIFIED_BORDER, TAB_LAST_PINNED_BORDER, TAB_SELECTED_BORDER_TOP, TAB_UNFOCUSED_ACTIVE_BACKGROUND, TAB_UNFOCUSED_ACTIVE_BORDER, TAB_UNFOCUSED_ACTIVE_BORDER_TOP, TAB_UNFOCUSED_ACTIVE_FOREGROUND, TAB_UNFOCUSED_ACTIVE_MODIFIED_BORDER, TAB_UNFOCUSED_HOVER_BACKGROUND, TAB_UNFOCUSED_HOVER_BORDER, TAB_UNFOCUSED_HOVER_FOREGROUND, TAB_UNFOCUSED_INACTIVE_BACKGROUND, TAB_UNFOCUSED_INACTIVE_MODIFIED_BORDER, TITLE_BAR_ACTIVE_BACKGROUND, TITLE_BAR_ACTIVE_FOREGROUND, TITLE_BAR_BORDER, TITLE_BAR_INACTIVE_BACKGROUND, TITLE_BAR_INACTIVE_FOREGROUND, WINDOW_ACTIVE_BORDER, WINDOW_INACTIVE_BORDER, WORKBENCH_BACKGROUND };
