// @codingame/monaco-vscode-localization-service-override/vscode/src/vs/workbench/contrib/localization/common/localizationsActions.js

import { localize2, localize } from '@codingame/monaco-vscode-api/vscode/vs/nls';
import { IQuickInputService } from '@codingame/monaco-vscode-api/vscode/vs/platform/quickinput/common/quickInput.service';
import { CancellationTokenSource } from '@codingame/monaco-vscode-api/vscode/vs/base/common/cancellation';
import { DisposableStore } from '@codingame/monaco-vscode-api/vscode/vs/base/common/lifecycle';
import { Action2, MenuId } from '@codingame/monaco-vscode-api/vscode/vs/platform/actions/common/actions';
import { ILanguagePackService } from '@codingame/monaco-vscode-api/vscode/vs/platform/languagePacks/common/languagePacks.service';
import { ILocaleService } from '@codingame/monaco-vscode-api/vscode/vs/workbench/services/localization/common/locale.service';
import { IExtensionsWorkbenchService } from '@codingame/monaco-vscode-api/vscode/vs/workbench/contrib/extensions/common/extensions.service';

class ConfigureDisplayLanguageAction extends Action2 {
    static { this.ID = 'workbench.action.configureLocale'; }
    constructor() {
        super({
            id: ConfigureDisplayLanguageAction.ID,
            title: ( localize2(7268, "Configure Display Language")),
            menu: {
                id: MenuId.CommandPalette
            },
            metadata: {
                description: ( localize2(
                    7269,
                    "Changes the locale of VS Code based on installed language packs. Common languages include French, Chinese, Spanish, Japanese, German, Korean, and more."
                ))
            }
        });
    }
    async run(accessor) {
        const languagePackService = accessor.get(ILanguagePackService);
        const quickInputService = accessor.get(IQuickInputService);
        const localeService = accessor.get(ILocaleService);
        const extensionWorkbenchService = accessor.get(IExtensionsWorkbenchService);
        const installedLanguages = await languagePackService.getInstalledLanguages();
        const disposables = ( new DisposableStore());
        const qp = disposables.add(quickInputService.createQuickPick({ useSeparators: true }));
        qp.matchOnDescription = true;
        qp.placeholder = ( localize(7270, "Select Display Language"));
        if (installedLanguages?.length) {
            const items = [{ type: 'separator', label: ( localize(7271, "Installed")) }];
            qp.items = items.concat(this.withMoreInfoButton(installedLanguages));
        }
        const source = ( new CancellationTokenSource());
        disposables.add(qp.onDispose(() => {
            source.cancel();
            disposables.dispose();
        }));
        const installedSet = ( new Set(installedLanguages?.map(language => language.id) ?? []));
        languagePackService.getAvailableLanguages().then(availableLanguages => {
            const newLanguages = availableLanguages.filter(l => l.id && !( installedSet.has(l.id)));
            if (newLanguages.length) {
                qp.items = [
                    ...qp.items,
                    { type: 'separator', label: ( localize(7272, "Available")) },
                    ...this.withMoreInfoButton(newLanguages)
                ];
            }
            qp.busy = false;
        });
        disposables.add(qp.onDidAccept(async () => {
            const selectedLanguage = qp.activeItems[0];
            if (selectedLanguage) {
                qp.hide();
                await localeService.setLocale(selectedLanguage);
            }
        }));
        disposables.add(qp.onDidTriggerItemButton(async (e) => {
            qp.hide();
            if (e.item.extensionId) {
                await extensionWorkbenchService.open(e.item.extensionId);
            }
        }));
        qp.show();
        qp.busy = true;
    }
    withMoreInfoButton(items) {
        for (const item of items) {
            if (item.extensionId) {
                item.buttons = [{
                        tooltip: ( localize(7273, "More Info")),
                        iconClass: 'codicon-info'
                    }];
            }
        }
        return items;
    }
}
class ClearDisplayLanguageAction extends Action2 {
    static { this.ID = 'workbench.action.clearLocalePreference'; }
    static { this.LABEL = ( localize2(7274, "Clear Display Language Preference")); }
    constructor() {
        super({
            id: ClearDisplayLanguageAction.ID,
            title: ClearDisplayLanguageAction.LABEL,
            menu: {
                id: MenuId.CommandPalette
            }
        });
    }
    async run(accessor) {
        const localeService = accessor.get(ILocaleService);
        await localeService.clearLocalePreference();
    }
}

export { ClearDisplayLanguageAction, ConfigureDisplayLanguageAction };
