// @codingame/monaco-vscode-api/vscode/src/vs/base/common/observableInternal/index.js

import '../arrays.js';
import '../event.js';
import '../lifecycle.js';
import { setLogObservableFn } from './logging/logging.js';
export { autorun, autorunDelta, autorunHandleChanges, autorunOpts, autorunWithStore, autorunWithStoreHandleChanges } from './autorun.js';
export { derived, derivedDisposable, derivedHandleChanges, derivedOpts, derivedWithSetter, derivedWithStore } from './derived.js';
import '../cancellation.js';
export { ValueWithChangeEventFromObservable, constObservable, derivedConstOnceDefined, derivedObservableWithCache, derivedObservableWithWritableCache, keepObserved, latestChangedValue, mapObservableArrayCached, observableFromEvent, observableFromEventOpts, observableFromPromise, observableFromValueWithChangeEvent, observableSignal, observableSignalFromEvent, recomputeInitiallyAndOnChange, runOnChange, runOnChangeWithStore, wasEventTriggeredRecently } from './utils.js';
import { logObservableToConsole } from './logging/consoleObservableLogger.js';

setLogObservableFn(logObservableToConsole);
