// @codingame/monaco-vscode-api/vscode/src/vs/workbench/services/extensionManagement/common/extensionManagement.js

import { FileAccess } from '../../../../base/common/network.js';

var ExtensionInstallLocation;
(function (ExtensionInstallLocation) {
    ExtensionInstallLocation[ExtensionInstallLocation["Local"] = 1] = "Local";
    ExtensionInstallLocation[ExtensionInstallLocation["Remote"] = 2] = "Remote";
    ExtensionInstallLocation[ExtensionInstallLocation["Web"] = 3] = "Web";
})(ExtensionInstallLocation || (ExtensionInstallLocation = {}));
const getDefaultIconPath = () => ( ( FileAccess.asBrowserUri('vs/workbench/services/extensionManagement/common/media/defaultIcon.png')).toString(true));
var EnablementState;
(function (EnablementState) {
    EnablementState[EnablementState["DisabledByTrustRequirement"] = 0] = "DisabledByTrustRequirement";
    EnablementState[EnablementState["DisabledByExtensionKind"] = 1] = "DisabledByExtensionKind";
    EnablementState[EnablementState["DisabledByEnvironment"] = 2] = "DisabledByEnvironment";
    EnablementState[EnablementState["EnabledByEnvironment"] = 3] = "EnabledByEnvironment";
    EnablementState[EnablementState["DisabledByVirtualWorkspace"] = 4] = "DisabledByVirtualWorkspace";
    EnablementState[EnablementState["DisabledByInvalidExtension"] = 5] = "DisabledByInvalidExtension";
    EnablementState[EnablementState["DisabledByAllowlist"] = 6] = "DisabledByAllowlist";
    EnablementState[EnablementState["DisabledByExtensionDependency"] = 7] = "DisabledByExtensionDependency";
    EnablementState[EnablementState["DisabledGlobally"] = 8] = "DisabledGlobally";
    EnablementState[EnablementState["DisabledWorkspace"] = 9] = "DisabledWorkspace";
    EnablementState[EnablementState["EnabledGlobally"] = 10] = "EnabledGlobally";
    EnablementState[EnablementState["EnabledWorkspace"] = 11] = "EnabledWorkspace";
})(EnablementState || (EnablementState = {}));

export { EnablementState, ExtensionInstallLocation, getDefaultIconPath };
