// @codingame/monaco-vscode-api/vscode/src/vs/editor/common/standaloneStrings.js

import { localize } from '../../nls.js';

var AccessibilityHelpNLS;
(function (AccessibilityHelpNLS) {
    AccessibilityHelpNLS.accessibilityHelpTitle = ( localize(790, "Accessibility Help"));
    AccessibilityHelpNLS.openingDocs = ( localize(791, "Opening the Accessibility documentation page."));
    AccessibilityHelpNLS.readonlyDiffEditor = ( localize(792, "You are in a read-only pane of a diff editor."));
    AccessibilityHelpNLS.editableDiffEditor = ( localize(793, "You are in a pane of a diff editor."));
    AccessibilityHelpNLS.readonlyEditor = ( localize(794, "You are in a read-only code editor."));
    AccessibilityHelpNLS.editableEditor = ( localize(795, "You are in a code editor."));
    AccessibilityHelpNLS.defaultWindowTitleIncludesEditorState = ( localize(
        796,
        "activeEditorState - such as modified, problems, and more, is included as a part of the window.title setting by default. Disable it with accessibility.windowTitleOptimized."
    ));
    AccessibilityHelpNLS.defaultWindowTitleExcludingEditorState = ( localize(
        797,
        "activeEditorState - such as modified, problems, and more, is currently not included as a part of the window.title setting by default. Enable it with accessibility.windowTitleOptimized."
    ));
    AccessibilityHelpNLS.toolbar = ( localize(
        798,
        "Around the workbench, when the screen reader announces you've landed in a toolbar, use narrow keys to navigate between the toolbar's actions."
    ));
    AccessibilityHelpNLS.changeConfigToOnMac = ( localize(
        799,
        "Configure the application to be optimized for usage with a Screen Reader (Command+E)."
    ));
    AccessibilityHelpNLS.changeConfigToOnWinLinux = ( localize(
        800,
        "Configure the application to be optimized for usage with a Screen Reader (Control+E)."
    ));
    AccessibilityHelpNLS.auto_on = ( localize(
        801,
        "The application is configured to be optimized for usage with a Screen Reader."
    ));
    AccessibilityHelpNLS.auto_off = ( localize(
        802,
        "The application is configured to never be optimized for usage with a Screen Reader."
    ));
    AccessibilityHelpNLS.screenReaderModeEnabled = ( localize(803, "Screen Reader Optimized Mode enabled."));
    AccessibilityHelpNLS.screenReaderModeDisabled = ( localize(804, "Screen Reader Optimized Mode disabled."));
    AccessibilityHelpNLS.tabFocusModeOnMsg = ( localize(
        805,
        "Pressing Tab in the current editor will move focus to the next focusable element. Toggle this behavior{0}.",
        '<keybinding:editor.action.toggleTabFocusMode>'
    ));
    AccessibilityHelpNLS.tabFocusModeOffMsg = ( localize(
        806,
        "Pressing Tab in the current editor will insert the tab character. Toggle this behavior{0}.",
        '<keybinding:editor.action.toggleTabFocusMode>'
    ));
    AccessibilityHelpNLS.stickScroll = ( localize(
        807,
        "Focus Sticky Scroll{0} to focus the currently nested scopes.",
        '<keybinding:editor.action.focusStickyDebugConsole>'
    ));
    AccessibilityHelpNLS.suggestActions = ( localize(
        808,
        "Trigger the suggest widget{0} to show possible code completions.",
        '<keybinding:editor.action.triggerSuggest>'
    ));
    AccessibilityHelpNLS.acceptSuggestAction = ( localize(
        809,
        "Accept suggestion{0} to accept the currently selected suggestion.",
        '<keybinding:acceptSelectedSuggestion>'
    ));
    AccessibilityHelpNLS.toggleSuggestionFocus = ( localize(
        810,
        "Toggle focus between the suggest widget and the editor{0} and toggle details focus with{1} to learn more about the suggestion.",
        '<keybinding:focusSuggestion>',
        '<keybinding:toggleSuggestionFocus>'
    ));
    AccessibilityHelpNLS.codeFolding = ( localize(
        811,
        "Use code folding to collapse blocks of code and focus on the code you're interested in via the Toggle Folding Command{0}.",
        '<keybinding:editor.toggleFold>'
    ));
    AccessibilityHelpNLS.intellisense = ( localize(
        812,
        "Use Intellisense to improve coding efficiency and reduce errors. Trigger suggestions{0}.",
        '<keybinding:editor.action.triggerSuggest>'
    ));
    AccessibilityHelpNLS.showOrFocusHover = ( localize(
        813,
        "Show or focus the hover{0} to read information about the current symbol.",
        '<keybinding:editor.action.showHover>'
    ));
    AccessibilityHelpNLS.goToSymbol = ( localize(
        814,
        "Go to Symbol{0} to quickly navigate between symbols in the current file.",
        '<keybinding:workbench.action.gotoSymbol>'
    ));
    AccessibilityHelpNLS.showAccessibilityHelpAction = ( localize(815, "Show Accessibility Help"));
    AccessibilityHelpNLS.listSignalSounds = ( localize(
        816,
        "Run the command: List Signal Sounds for an overview of all sounds and their current status."
    ));
    AccessibilityHelpNLS.listAlerts = ( localize(
        817,
        "Run the command: List Signal Announcements for an overview of announcements and their current status."
    ));
    AccessibilityHelpNLS.quickChat = ( localize(
        818,
        "Toggle quick chat{0} to open or close a chat session.",
        '<keybinding:workbench.action.quickchat.toggle>'
    ));
    AccessibilityHelpNLS.startInlineChat = ( localize(
        819,
        "Start inline chat{0} to create an in editor chat session.",
        '<keybinding:inlineChat.start>'
    ));
    AccessibilityHelpNLS.startDebugging = ( localize(
        820,
        "The Debug: Start Debugging command{0} will start a debug session.",
        '<keybinding:workbench.action.debug.start>'
    ));
    AccessibilityHelpNLS.setBreakpoint = ( localize(
        821,
        "The Debug: Inline Breakpoint command{0} will set or unset a breakpoint at the current cursor position in the active editor.",
        '<keybinding:editor.debug.action.toggleInlineBreakpoint>'
    ));
    AccessibilityHelpNLS.addToWatch = ( localize(
        822,
        "The Debug: Add to Watch command{0} will add the selected text to the watch view.",
        '<keybinding:editor.debug.action.selectionToWatch>'
    ));
    AccessibilityHelpNLS.debugExecuteSelection = ( localize(
        823,
        "The Debug: Execute Selection command{0} will execute the selected text in the debug console.",
        '<keybinding:editor.debug.action.selectionToRepl>'
    ));
    AccessibilityHelpNLS.chatEditorModification = ( localize(
        824,
        "The editor contains pending modifications that have been made by chat."
    ));
    AccessibilityHelpNLS.chatEditorRequestInProgress = ( localize(
        825,
        "The editor is currently waiting for modifications to be made by chat."
    ));
    AccessibilityHelpNLS.chatEditActions = ( localize(
        826,
        'Navigate between edits in the editor with navigate previous{0} and next{1} and accept{2}, reject{3} or view the diff{4} for the current change.',
        '<keybinding:chatEditor.action.navigatePrevious>',
        '<keybinding:chatEditor.action.navigateNext>',
        '<keybinding:chatEditor.action.acceptHunk>',
        '<keybinding:chatEditor.action.undoHunk>',
        '<keybinding:chatEditor.action.toggleDiff>'
    ));
})(AccessibilityHelpNLS || (AccessibilityHelpNLS = {}));
var InspectTokensNLS;
(function (InspectTokensNLS) {
    InspectTokensNLS.inspectTokensAction = ( localize(827, "Developer: Inspect Tokens"));
})(InspectTokensNLS || (InspectTokensNLS = {}));
var GoToLineNLS;
(function (GoToLineNLS) {
    GoToLineNLS.gotoLineActionLabel = ( localize(828, "Go to Line/Column..."));
})(GoToLineNLS || (GoToLineNLS = {}));
var QuickHelpNLS;
(function (QuickHelpNLS) {
    QuickHelpNLS.helpQuickAccessActionLabel = ( localize(829, "Show all Quick Access Providers"));
})(QuickHelpNLS || (QuickHelpNLS = {}));
var QuickCommandNLS;
(function (QuickCommandNLS) {
    QuickCommandNLS.quickCommandActionLabel = ( localize(830, "Command Palette"));
    QuickCommandNLS.quickCommandHelp = ( localize(831, "Show And Run Commands"));
})(QuickCommandNLS || (QuickCommandNLS = {}));
var QuickOutlineNLS;
(function (QuickOutlineNLS) {
    QuickOutlineNLS.quickOutlineActionLabel = ( localize(832, "Go to Symbol..."));
    QuickOutlineNLS.quickOutlineByCategoryActionLabel = ( localize(833, "Go to Symbol by Category..."));
})(QuickOutlineNLS || (QuickOutlineNLS = {}));
var StandaloneCodeEditorNLS;
(function (StandaloneCodeEditorNLS) {
    StandaloneCodeEditorNLS.editorViewAccessibleLabel = ( localize(834, "Editor content"));
})(StandaloneCodeEditorNLS || (StandaloneCodeEditorNLS = {}));
var ToggleHighContrastNLS;
(function (ToggleHighContrastNLS) {
    ToggleHighContrastNLS.toggleHighContrast = ( localize(835, "Toggle High Contrast Theme"));
})(ToggleHighContrastNLS || (ToggleHighContrastNLS = {}));
var StandaloneServicesNLS;
(function (StandaloneServicesNLS) {
    StandaloneServicesNLS.bulkEditServiceSummary = ( localize(836, "Made {0} edits in {1} files"));
})(StandaloneServicesNLS || (StandaloneServicesNLS = {}));

export { AccessibilityHelpNLS, GoToLineNLS, InspectTokensNLS, QuickCommandNLS, QuickHelpNLS, QuickOutlineNLS, StandaloneCodeEditorNLS, StandaloneServicesNLS, ToggleHighContrastNLS };
